File: Pylon/data/viewer/utils/displays/aabbs/twod/ts/backend/apis.py
"""2D axis-aligned-box display response APIs."""

from typing import List, Optional

from data.viewer.utils.displays.aabbs.twod.ts.backend.schemas.display_response import (
    Aabb2dDisplayResponse,
)


def create_aabb_2d_display_response(
    slot_id: str,
    title: str,
    aabbs: List[List[float]],
    scores: Optional[List[float]] = None,
) -> Aabb2dDisplayResponse:
    """Create a 2D axis-aligned-box overlay display response.

    Args:
        slot_id: Stable display slot identifier.
        title: Display panel title.
        aabbs: Inline axis-aligned 2D boxes, each a 4-float list
            ``[min_x, min_y, max_x, max_y]`` in image-pixel coordinates.
        scores: Optional per-box scores, one per box in ``aabbs``.

    Returns:
        2D axis-aligned-box overlay display response.
    """

    def _validate_inputs() -> None:
        assert isinstance(slot_id, str), (
            "Slot id must be a string. slot_id=%r" % slot_id
        )
        assert isinstance(title, str), "Title must be a string. title=%r" % title
        assert isinstance(aabbs, list), "Boxes must be a list. aabbs=%r" % aabbs
        for box in aabbs:
            assert isinstance(box, list), "Each box must be a list. box=%r" % box
            assert len(box) == 4, (
                "Each 2D box must have 4 floats [min_x, min_y, max_x, max_y]. box=%r"
                % box
            )
            for coordinate in box:
                assert isinstance(coordinate, float), (
                    "Each box coordinate must be a float. coordinate=%r" % coordinate
                )
        assert scores is None or isinstance(scores, list), (
            "Scores must be a list or None. scores=%r" % scores
        )
        if scores is not None:
            assert len(scores) == len(
                aabbs
            ), "Scores must have one entry per box. len(scores)=%d len(aabbs)=%d" % (
                len(scores),
                len(aabbs),
            )
            for score in scores:
                assert isinstance(score, float), (
                    "Each score must be a float. score=%r" % score
                )

    _validate_inputs()

    return Aabb2dDisplayResponse(
        slot_id=slot_id,
        title=title,
        aabbs=aabbs,
        scores=scores,
    )

File: Pylon/data/viewer/utils/displays/aabbs/twod/ts/frontend/apis.ts
import type { LeafVNode } from "web/reconcile/reconcile";
import {
  registerRasterLayerRenderer,
  type RasterLayerRenderer,
} from "data/viewer/utils/displays/utils/ts/frontend/layer_renderer_registry";
import type { Aabb2dDisplayResponse } from "./types/display_response";

const SVG_NS = "http://www.w3.org/2000/svg";

// Box stroke color and width, and score-label style for the 2D box overlay.
const AABB_2D_BOX_STROKE = "#4da6ff";
const AABB_2D_BOX_STROKE_WIDTH = "2";
const AABB_2D_LABEL_FONT_SIZE = "14";

// Renders the inline 2D axis-aligned boxes and their optional per-box score
// labels as a raster overlay layer over an image frame: a LeafVNode whose
// absolutely-positioned full-bleed SVG draws each box as a <rect> with its score
// <text>, so a consumer composes it over an image as a raster aux layer. The SVG
// is full-bleed with preserveAspectRatio="none" and its <rect>s carry base-image
// pixel coordinates; the layered container owns the viewBox, setting it to the
// shared raster frustum on the base image's load.
//
// Args:
//   displayResponse: the 2D box overlay response carrying the inline boxes (each
//     [min_x, min_y, max_x, max_y], image-pixel coordinates) and the optional
//     per-box scores.
//
// Returns:
//   A LeafVNode whose render() mounts the raster box overlay.
export function renderAabb2dDisplay({
  displayResponse,
}: {
  displayResponse: Aabb2dDisplayResponse;
}): LeafVNode {
  const leaf: LeafVNode = {
    kind: "leaf",
    key: displayResponse.url ?? `aabb_2d:${displayResponse.slot_id}`,
    props: {},
    render: () => _buildBoxesOverlay({ displayResponse }),
  };
  return leaf;
}

// Build the box overlay element: an absolutely-positioned full-bleed div hosting
// a full-bleed SVG (preserveAspectRatio="none"), drawing each box as a stroke
// <rect> and its score (when present) as a <text> just above the box, in
// base-image pixel coordinates. The layered container owns the SVG viewBox.
//
// Args:
//   displayResponse: the 2D box overlay response carrying the boxes and optional
//     per-box scores.
//
// Returns:
//   The overlay div hosting the box/score SVG.
function _buildBoxesOverlay({
  displayResponse,
}: {
  displayResponse: Aabb2dDisplayResponse;
}): HTMLElement {
  const boxes = displayResponse.aabbs;
  const scores = displayResponse.scores;
  const overlay = document.createElement("div");
  overlay.style.position = "absolute";
  overlay.style.inset = "0";
  overlay.style.pointerEvents = "none";
  const svg = document.createElementNS(SVG_NS, "svg");
  svg.setAttribute("preserveAspectRatio", "none");
  svg.style.position = "absolute";
  svg.style.inset = "0";
  svg.style.width = "100%";
  svg.style.height = "100%";
  svg.style.pointerEvents = "none";
  overlay.append(svg);
  for (let boxIndex = 0; boxIndex < boxes.length; boxIndex += 1) {
    const [x1, y1, x2, y2] = boxes[boxIndex];
    const rect = document.createElementNS(SVG_NS, "rect");
    rect.setAttribute("x", String(x1));
    rect.setAttribute("y", String(y1));
    rect.setAttribute("width", String(x2 - x1));
    rect.setAttribute("height", String(y2 - y1));
    rect.setAttribute("fill", "none");
    rect.setAttribute("stroke", AABB_2D_BOX_STROKE);
    rect.setAttribute("stroke-width", AABB_2D_BOX_STROKE_WIDTH);
    svg.append(rect);
    if (scores !== null) {
      const label = document.createElementNS(SVG_NS, "text");
      label.setAttribute("x", String(x1));
      label.setAttribute("y", String(Math.max(y1 - 3, 10)));
      label.setAttribute("fill", AABB_2D_BOX_STROKE);
      label.setAttribute("font-size", AABB_2D_LABEL_FONT_SIZE);
      label.setAttribute("font-family", "monospace");
      label.textContent = scores[boxIndex].toFixed(2);
      svg.append(label);
    }
  }
  return overlay;
}

// Module-load self-registration of the raster aabb-2d layer renderer. The
// registry erases the layer's display response to the base DisplayResponse, so
// the aabb-2d part-B is registered through the registry's renderer type at the
// erasure boundary.
registerRasterLayerRenderer({
  displayKind: "aabb_2d",
  layerRenderer: renderAabb2dDisplay as RasterLayerRenderer,
});
